src/pudl/transform/vceregen.py

```diff
@@ -0,0 +1,206 @@
+
+    This function adds a datetime column and a year_hour column.
+    The datetime column is important for merging the data with other
+    data, and the year_hour 0-8769 column is important for modeling
+    purposes. The report_year column is also helpful for filtering,
+    so we keep all three!
+
+    For leap years (2020), December 31st is excluded.
+
+    TO-DO: decide whether to have the year_hour start at 0 or 1.
+    and update the date_range accordingly.
+    """
+    logger.info("Adding datetime columns")
+    # This data is compiled for modeling purposes and skips the last
+    # day of a leap year. If we want to add a datetime column, we need
+    # to make sure that we skip the 31st of December, 2020.
+    # the [:-1] is because datetime will include both the 0th hour of the
+    # start date and the 0th hour of the end date. We only want one!
+    # Though we still need to decide which. This at least makes the math
+    # right.
+    to_leap_year_dates = pd.date_range(start="2019-01-01", end="2020-12-31", freq="h")[
+        :-1
+    ]
+    post_leap_year_dates = pd.date_range(
+        start="2021-01-01", end="2024-01-01", freq="h"
+    )[:-1]
+    # Add date sequence to pre and post leap year dates
+    df.loc[:, "report_year"] = df.report_year.astype("Int32")
+    df.loc[df["report_year"] < 2021, "hour"] = to_leap_year_dates
+    df.loc[df["report_year"] > 2020, "hour"] = post_leap_year_dates
+    # Add a year_date column for modeling purposes.
+    # Right now each year should go from 0-8759. We
+    # can decide to make that 1-8760 if we want to!
+    df.loc[:, "year_hour"] = df.hour.dt.hour + (df.hour.dt.dayofyear - 1) * 24
+    return df
+
+
+def _stack_cap_fac_df(df, df_name: pd.DataFrame) -> pd.DataFrame:
+    """Funciton to transform each capacity factor table individually to save memory.
+
+    The main transforms are turning county columns into county rows and renaming columns
+    to be more human-readable and compatible with the FIPs df that will get merged in.
+
+    This function is intended to save memory by being applied to each individual
+    capacity factor table rather than the giant combined one.
+    """
+    logger.info("Stacking the county columns")
+    df_stacked = (
+        df.set_index(["hour", "year_hour", "report_year"])
+        .stack()
+        .reset_index()
+        .rename(
+            columns={"level_3": "county_state_names", 0: f"capacity_factor_{df_name}"}
+        )
+        .assign(county_state_names=lambda x: pd.Categorical(x.county_state_names))
+    )
+    return df_stacked
+
+
+def _combine_all_cap_fac_dfs(cap_fac_dict: dict[str, pd.DataFrame]) -> pd.DataFrame:
+    """Combine capacity factor tables."""
+    logger.info("Merging all the capacity factor tables into one")
+    merge_keys = ["report_year", "hour", "year_hour", "county_state_names"]
+    # Can't merge all at once using reduce because that's too much memory
+    # So we'll merge one at a time.
+    mega_df = pd.merge(
```

Comment:
If you set the merge columns as the index you can use [`pd.concat()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to combine them, which I think should be less resource intensive than `pd.merge()` especially if they're all sorted by index first.